include/nbl/builtin/hlsl/fft/common.hlsl

```diff
@@ -1,60 +1,128 @@
 #ifndef _NBL_BUILTIN_HLSL_FFT_COMMON_INCLUDED_
 #define _NBL_BUILTIN_HLSL_FFT_COMMON_INCLUDED_
 
-#include "nbl/builtin/hlsl/complex.hlsl"
-#include "nbl/builtin/hlsl/cpp_compat.hlsl"
+#include <nbl/builtin/hlsl/cpp_compat.hlsl>
+#include <nbl/builtin/hlsl/complex.hlsl>
+#include <nbl/builtin/hlsl/concepts.hlsl>
+#include <nbl/builtin/hlsl/math/intutil.hlsl>
 #include "nbl/builtin/hlsl/numbers.hlsl"
+#include "nbl/builtin/hlsl/concepts.hlsl"
 
-namespace nbl 
+namespace nbl
 {
 namespace hlsl
 {
-namespace fft 
+namespace fft
 {
 
+// template parameter N controls the number of dimensions of the input
+// template parameter M controls the number of dimensions to pad up to PoT
+// "axes" indicates which dimensions to pad up to PoT
+template <uint16_t N, uint16_t M NBL_FUNC_REQUIRES(M <= N)
+inline vector<uint64_t, 3> padDimensions(NBL_CONST_REF_ARG(vector<uint32_t, N>) dimensions, NBL_CONST_REF_ARG(vector<uint16_t, M>) axes, bool realFFT = false)
+{
+    vector<uint32_t, N> newDimensions = dimensions;
+    uint16_t axisCount = 0;
+    for (uint16_t i = 0u; i < M; i++)
+    {
+        newDimensions[i] = hlsl::roundUpToPoT(newDimensions[i]);
+        if (realFFT && !axisCount++)
+            newDimensions[i] /= 2;
+    }
+    return newDimensions;
+}
+
+// template parameter N controls the number of dimensions of the input
+// template parameter M controls the number of dimensions we run an FFT along AND store the result
+// "axes" indicates which dimensions we run an FFT along AND store the result
+template <uint16_t N, uint16_t M NBL_FUNC_REQUIRES(M <= N)
+inline uint64_t getOutputBufferSize(NBL_CONST_REF_ARG(vector<uint32_t, N>) inputDimensions, uint32_t numChannels, NBL_CONST_REF_ARG(vector<uint16_t, M>) axes, bool realFFT = false, bool halfFloats = false)
+{
+    const vector<uint64_t, 3> paddedDims = padDimensions<N, M>(inputDimensions, axes);
```

Comment:
the axes stuff is mightly confusing, see https://github.com/Devsh-Graphics-Programming/Nabla/pull/795#discussion_r1888150303

Reply:
The thing is you only want to pad along the dimensions you're going to run an FFT and store the dimensions afterward. For example for 2D FFT running y-axis first you only want to pad the y-axis since you do FFT along y -> store -> FFT along x -> NO store -> conv + IFFT along x -> only then store back

Reply:
If you want the full result of a 2D FFT only then you pad both dimensions

Reply:
hmm, ok I see your point, problem was that your `axes` variable was unused in `padDimensions`

I'd always have `padDimensions` return the full final FFT size (for this only matters if the FFT is of a real signal and if it is what axis goes first)

Then `outputBufferSize` can return the buffer size after a certain pass.

For the stuff to make sense semantically, instead of `vector<uint16_t,N> axes` you need to call it `passAxisOrder` (component 0 tells you along which basis we perform the FFT), and default would be (0,1,2) (a.k.a. `iota`)

It makes no sense to have `N` and `M`, just `N` would do.

You'd also need a `const uint16_t pass`. so 
```cpp
uint64_t getOutputBufferSize(
   const uint32_t numChannels
   const vector<uint32_t,N> inputDimensions,
   const uint16_t passIx,
   const vector<uint16_t,N> axisPassOrder = truncate<vector<uint16_t,N> >(uint16_t4(0,1,2,3)),
   bool realFFT = false,
   bool halfFloats = false
)
```

Alos the thing has so many arguments, it should probably be a struct with `operator()` instead XD

Although then you get no default params.

Reply:
> I'd always have padDimensions return the full final FFT size (for this only matters if the FFT is of a real signal and if it is what axis goes first)

Btw its easy and makes more sense to later just use `mix/lerp` to get the actualy output size:
```cpp
vector<bool,3> axesDone;
uint32_t3 passOutputDimension = mix(inputDimensions,paddedDimensions,axesDone);
```